### ISSUE TITLE
Feature suggestions for UX/UI and Eyecandy.

### ISSUE THREAD

**Issue by Kormann** [yquake2/yquake2#672]
Hello yamagi team. Here are some suggestions to make your q2port the ultimate vanilla experience not only for single player but for multiplayer:
1) Is it in your roadmap to add in-game server browser like Quake 3 / Q2Pro ? It would make the game more user friendly for newcomers to play multiplayer.
2) Scriptable menus like Q2Pro are nice if you want to customize the menu to fits better a mod, like mission packs, action, ctf, etc. It would be cool for teamplay menus as well like to add specific things in UI to like binding a key to drop a weapon and a ammo + a message for teamplay, etc.
3) Add a mods menu (q3 style), if it recognizes a folder with a game.dll it would appear in a list in the main menu. It would be easy for noobs to play a mod with zero effort, commands and batch files. The art would be easy to create, just create the word MODS using the letters from current Q2 vanilla main menu plaques.
4) Add the underwater warp effect like the original software one for all the renders. VKQuake2 did it and it's a nice addition to the original Q2 vanilla eyecandy experience.

Best Regards.

**Comment by Yamagi** [comment]
Hi :)

> Is it in your roadmap to add in-game server browser like Quake 3 / Q2Pro ? It would make the game more user friendly for newcomers to play multiplayer.

We considered adding a server browser and decided against it. At a first look it sounds easy. The server list is just a text file, that can be downloaded. curl is already part of YQ2. After that parse the file and feed the data in some kind of menu. But that text file contains arbitrary data that  can't be trusted and Q2s code is rather fragile. To implement the parsing in a safe manner that rules out triggering buffer overflows and other security problems somewhere else in the code is very hard. I don't have the time for that.

> Scriptable menus like Q2Pro are nice if you want to customize the menu to fits better a mod, like mission packs, action, ctf, etc. It would be cool for teamplay menus as well like to add specific things in UI to like binding a key to drop a weapon and a ammo + a message for teamplay, etc.

This is more or less the same. There's still the comment in the menu code saying that a better implementation would be nice. But designing and coding a menu that's scriptable, safe and backward compatible with the current stuff is hard.

That said, if someone steps up to do the work, walk the long way to rule all the corner cases and is willing to maintain the code in the long run, be my guest. I'm happy to merge pull requests. But neither @DanielGibson nor myself have the time to do it.

> Add the underwater warp effect like the original software one for all the renders. VKQuake2 did it and it's a nice addition to the original Q2 vanilla eyecandy experience.

This is indeed planned. The soft and vk renderers already have it, gl3 will get it some day. In gl1 it's impossible, since OpenGL 1.4 doesn't have shaders. The bug for gl3 is #514.

**Comment by Kormann** [comment]
@Yamagi is the Q2PRO Server browser not safe? Would it be unsafe to adapt it's code?

**Comment by Yamagi** [comment]
I haven't looked into the q2pro code an don't know. The q2pro guys usually know what they're doing, so I would assume that their code is okay. But q2pro diverged a lot from Vanilla Quake II and other source ports. Porting code from q2pro is often nit that easy, it needs some adoption for other codebases.

**Comment by apartfromtime** [comment]
> I haven't looked into the q2pro code an don't know. The q2pro guys usually know what they're doing, so I would assume that their code is okay. But q2pro diverged a lot from Vanilla Quake II and other source ports. Porting code from q2pro is often nit that easy, it needs some adoption for other codebases.

I'm working on some changes to bring the yq2 ui inline with that of q3. I can look into porting the q2pro ui to yq2, but as mentioned, q2pro has diverged a lot.

**Comment by Yamagi** [comment]
A good pull reuqest would be welcome. The big question is, if an ingame server browser is really worth the effort.

**Comment by apartfromtime** [comment]
Yes good question. I'm not convinced that scripting is a good idea either.

**Comment by apartfromtime** [comment]
I've made a scan through the code, apart from a general tidy of the menu code, rewriting everything similar to q3 team arena as commented in the code would require special cases for the menu cursor as q3a added the use of the mouse cursor and highlighting in lieu of the menu cursor. That wouldn't be vanilla, which is may be a reason no one has implemented the changes.
